src/ossl/ecdh.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright 2023 - 2024 Simo Sorce, Jakub Jelen
+// See LICENSE.txt file for terms
+
+use core::ffi::{c_char, c_int, c_uint};
+use std::borrow::Cow;
+
+use crate::attribute::CkAttrs;
+use crate::bytes_to_vec;
+use crate::ecc::*;
+use crate::error::Result;
+use crate::interface::*;
+use crate::mechanism::*;
+use crate::object::{default_key_attributes, Object, ObjectFactories};
+use crate::ossl::bindings::*;
+use crate::ossl::common::*;
+#[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
+use crate::ossl::ec_montgomery::{
+    is_montgomery_curve, make_ec_montgomery_public_key,
+    make_output_length_from_montgomery_obj,
+    montgomery_object_to_ecc_private_key,
+};
+#[cfg(feature = "ecc")]
+use crate::ossl::ecc::{
+    ecdsa_object_to_ecc_private_key, get_curve_name_from_obj,
+    make_ecdsa_public_key, make_output_length_from_ecdsa_obj,
+};
+
+fn kdf_type_to_hash_mech(mech: CK_EC_KDF_TYPE) -> Result<CK_MECHANISM_TYPE> {
+    match mech {
+        CKD_SHA1_KDF => Ok(CKM_SHA_1),
+        CKD_SHA224_KDF => Ok(CKM_SHA224),
+        CKD_SHA256_KDF => Ok(CKM_SHA256),
+        CKD_SHA384_KDF => Ok(CKM_SHA384),
+        CKD_SHA512_KDF => Ok(CKM_SHA512),
+        CKD_SHA3_224_KDF => Ok(CKM_SHA3_224),
+        CKD_SHA3_256_KDF => Ok(CKM_SHA3_256),
+        CKD_SHA3_384_KDF => Ok(CKM_SHA3_384),
+        CKD_SHA3_512_KDF => Ok(CKM_SHA3_512),
+        _ => return Err(CKR_MECHANISM_PARAM_INVALID)?,
+    }
+}
+
+fn make_derive_output_length_from_obj(key: &Object) -> Result<usize> {
+    // Montgomery
+    #[cfg(all(feature = "ec_montgomery", not(feature = "fips")))]
```

Comment:
> These functions should probably be exported by the relative mechanism and the mechanism should be sourced via the key type.

The problem is the mechanism is the same for both, but only key type differs.

> We may need to add a new trait to key objects that can be used to extract this info in a more abstract way.
Ie key.get_key_len() or key.get_field_len() or key.get_modulo_len() or some such.

This would be good, but I think it would require quite a lot of rework how the objects are working. For now, they are mostly abstract list of attributes, but to have this ability to "say" how large output will be or to "convert themselves" to Pkey, they would need more tying to the mechanism implementation and it is currently not clear to me how to do this. So I will open a follow-up issue.

Reply:
We actually already have something like this, which I introduced for symmetric keys, see impl SecretKeyFactory for AesKeyFactory

That said, I am ok deferring this to a separate ticket and address it later as a refactoring.

Reply:
Thanks for pointer! I probably missed that. Let me have a look if I will be able to figure out a better way. Rather do it right away than getting back to that in indefinite future.

Reply:
I am fine waiting, in retrospect I in fact prefer a separate PR to handle this issue, so I can fully dedicate my attention to a problem at a time in reviewing.

Reply:
Ok then. I will squash current changes and then it is ready. Will open follow-up PR with these changes once ready.